Enhance Makefile for Improved Service Management, Cleanup, and Utility Commands

liocle opened the pull request:
This pull request adds functionalities and user-friendliness to the Makefile. The key updates are:

1. **Service Management**:
    - Added `up_nginx`, `up_backend`, `up_db`, and `up_elk` targets for building and starting services individually.
    - Updated `up_no_elk` and `up_all` targets to rebuild images before starting services with the `--build` flag.

2. **Cleanup Enhancements**:
    - Introduced `clean_volumes`, `clean_orphans`, and `clean_images` targets for specific cleanup tasks.
    - Added `clean_all` target for a full cleanup of containers, networks, volumes, and images.

3. **Utility Commands**:
    - Added `logs_nginx`, `logs_backend`, `logs_db`, and `logs_elk` targets to tail logs from specific services.
    - Introduced `status` target to check the status of running containers.
    - Added `exec_nginx`, `exec_backend`, `exec_db`, and `exec_elk` targets to open a shell in specific containers.
    - Added `stop_all` target to stop all running services without removing them.

4. **Help Command**:
    - Implemented a `help` target to display available make targets and their descriptions.

These updates streamline development and debugging by providing better control over the Docker environment, quick access to logs, and container status.

**Benefits**:
- Improved clarity and usability of the Makefile.
- Simplified service management with specific targets for each service.
- Enhanced control over the Docker environment with comprehensive clean options and utility commands.

**Expected Outcome**:
A more robust and user-friendly Makefile that facilitates efficient service management, cleanup, and utility access during development and debugging.

**Related Issue**:
Closes #33

liocle committed:
Enhance service management and cleanup in Makefile

djames9 commented on Makefile:
```diff
@@ -1,25 +1,125 @@
+clean: clean_volumes clean_orphans
+
+# Rebuild and start all services
+re: clean_all up_no_elk
```
it needs to be modified make re, because now you delete the volumes but it should not be delete it. por favor trabaja bien !

liocle replied:
This issue has been solved.

linhtng commented on Makefile:
```diff
@@ -1,25 +1,125 @@
+	docker-compose -f ./docker-compose.yml logs -f es01 kibana logstash01 filebeat01 metricbeat01
+
+# Check status of running containers
+status:
```
can we pls have something shorter, like just ps? Because typing make ps faster than docker ps faster than make status. So, ps or stt

liocle replied:
This issue has been solved.

liocle committed:
Fix Makefile issues and add grouped functionalities

liocle commented:
Hi @linhtng @djames9 @pixelsnow  @wengcychan here is a revised version of the updated Makefile.

# Summary of Makefile Changes

## Overview
This update involves reorganizing and enhancing the Makefile for our project. The primary goals were to improve readability, ensure ease of use, and add valuable functionalities for log management and container status inspection.

## Changes Made

### 1. Grouping by Functionality
The Makefile targets are now grouped by their functionalities to improve readability and make it easier for users to find the commands they need. The groups are:
- **Build and Start**
- **Clean and Remove**
- **Rebuild and Restart**
- **Logging**
- **Status**
- **Exec into Containers**
- **Stop Services**
- **Monitoring**
- **Help**

### 2. New Targets Added
Several new targets have been added to provide more functionality:
- **Logs:**
  - `logs_errors`: Tail logs with timestamps and filter for `ERROR` messages.
  - `logs_grep`: Tail logs with timestamps and filter by a user-specified keyword.
- **Status:**
  - `ps_short`: Check status of running containers with concise output.
  - `ps_inspect`: Inspect a specific container, combining `ps_short` for concise status and then prompting for the container name to inspect.
- **Monitoring:**
  - `stats`: Show resource utilization statistics for running containers.
  - `sys_df`: Show Docker system disk usage.

### 3. Consistent Naming and Functionalities
All targets have been named consistently, and their functionalities are clearly defined. For example:
- `up_*` targets for starting services.
- `clean_*` targets for cleaning up specific resources.
- `re_*` targets for rebuilding and restarting services.

### 4. Updated Help Section
The `help` target now provides a clear and organized summary of all available commands, grouped by their functionalities. This makes it easier for users to quickly find the command they need.

Please review the updated Makefile and provide any further feedback.

### 5. Related Issue:
Closes #33

linhtng commented on Makefile:
```diff
@@ -55,9 +59,24 @@ clean_all: clean_volumes clean_orphans clean_images
+# Rebuild and Restart
+################################################################################
+# Rebuild and start all services without removing volumes
+re: clean_orphans clean_images up_no_elk
```
The combination of 
```
docker-compose -f ./docker-compose.yml down --remove-orphans
docker rmi -f $$(docker images -aq) 
```
is rather inconsistent, afaik. The 1st command will only stop and remove containers, networks, and orphan containers created by the specific docker-compose.yml file you specify, which is the one in the same directory. However, the 2nd command will attempt to remove all Docker images on your system, regardless of which yaml file or directory they were created from. This can indeed remove images that are being used by other containers not managed by the current yaml file (thus, still up and running). 
So, we should change one of them to be consistent with the other, maybe change the 2nd one so that it will only remove the imgs created by current yaml file, I think.

liocle replied:
Fixing this was trickier than I thought, but since this target rule was needed, it was important to control what images can be removed. For that, I had to create a temporary file to store image IDs and keep track of it's content until their deletion. Thanks for your eagle eyes at spotting these issues, let's see if you find some more 🐛🔭🦅🙏

linhtng commented on Makefile:
@liocle I'm so sorry that I didn't notice it yesterday that we are now using "docker-compose" instead of "docker compose" for all the commands. I remember Markus told me specifically about this that I should use "docker compose" because it is up-to-date, new version instead of the outdated version "docker-compose".

liocle replied:
Sorry, I had not seen this issue earlier,  it is now addressed. Please @wengcychan @pixelsnow  @djames9  test it well on your side with the functionalities useful to you. The makefile seems fine now, but I have every now and then some odd behaviors. I do not believe that the makefile is destructive to our project or prevent it to run, so I could also fix it after merge if needs be. Though, beware to the new  git nuke rule I have added this one *is* destructive (courtesy of Markus).

djames9 replied:
I review it , but "make re" is not working properly. it does one extra docker compose down and it does not initialize the project.

djames9 replied:
I am sorry I did not see this before but the build_all  it does not work maybe we can take it out, or what do you thing? and maybe up_nginx is redundant

liocle replied:
I agree and removed the 'build_all' and 'up_nginx' that were either not found useful/redundant. Thanks for the review!

djames9 replied:
I am sorry . I did not see this before but I think " make  all            Build and start all services except ELK stack
  up_no_elk      Build and start all services except ELK stack". I think this is redundant. SO if we only keep one that will be great

liocle replied:
Sorry, but 'all' depends on 'up_no_elk', so I wont remove it, however, I can remove the line in the make help to not clutter it.

liocle committed:
Fix deletion of non project related containers and images

liocle committed:
change deprecated docker-compose into docker compose

liocle committed:
Rename pre_clean_all target into get_images_id and add git nuke functionality

git nuke removes all untracked files and directories from ther working directory.
It resets the working directory and the index to the last commit, discarding all changes to tracked files.

liocle committed:
Fix 'make re' not working

liocle committed:
Remove build_all and up_nginx target rules

'build_all' was not working properly and not found useful as functionality.

'up_ngin' was found to be a replicate of 'all' and extending the Makefile unecessarily.

liocle committed:
Remove redundent up_no_helk from 'make help'

linhtng reviewed (approved):
Looks good to me.
Great work, Lionel!🌟
